lib/db/schemaupdater.go

```diff
@@ -683,6 +685,84 @@ func (db *schemaUpdater) updateSchemaTo13(prev int) error {
 	return t.Commit()
 }
 
+func (db *schemaUpdater) updateSchemaTo14(_ int) error {
```

Comment:
This seems overly complicated, isn't it just rewriteFiles plus a GC pass?

Reply:
That doesn't cover `BlockListMap`. However I just realized there's an alternative option that doesn't require a migration and addresses the conflict problem. It is a bit less efficient but more robust, as it also prevents conflicts e.g. between a device on an RC (already done the migration) and one on stable:

```diff
--- a/lib/protocol/bep_extensions.go
+++ b/lib/protocol/bep_extensions.go
@@ -276,9 +276,11 @@ func PermsEqual(a, b uint32) bool {
 
 // BlocksEqual returns true when the two files have identical block lists.
 func (f FileInfo) BlocksEqual(other FileInfo) bool {
-       // If both sides have blocks hashes then we can just compare those.
-       if len(f.BlocksHash) > 0 && len(other.BlocksHash) > 0 {
-               return bytes.Equal(f.BlocksHash, other.BlocksHash)
+       // If both sides have blocks hashes and they match, we are good. If they
+       // don't match still check individual block hashes to catch differences
+       // in weak hashes only (e.g. after switching weak hash algo).
+       if len(f.BlocksHash) > 0 && len(other.BlocksHash) > 0 && bytes.Equal(f.BlocksHash, other.BlocksHash) {
+               return true
        }
 
        // Actually compare the block lists in full.
```

Any gotchas coming to mind going that route?

Reply:
Seems fine to me